Issue 2694 update schema design docs (#3215)
IBM/FHIR: fhir-persistence-schema/docs/SchemaDesign.md

# The IBM FHIR Server - Schema Design and Management

_Note_: the following description is based on the standard Derby/PostgreSQL variant of the schema. For Db2, the logical model is the same, but the physical model uses table partitioning and row-based access control (RBAC) to support multi-tenancy. To facilitate this, all the primary and foreign key relationships are prefixed with a multi-tenant identifier `mt_id`. A complete description of the Db2 multi-tenant design can be found [here](DB2MultiTenancy.md). Other variations across databases are discussed in this document.

> Currently it is not possible to deploy the simplified single-tenant variant of the schema to Db2. GitHub [issue-840](https://github.com/IBM/FHIR/issues/840) has been opened to address this. 

Reviewer: ```suggestion
Currently it is not possible to deploy the single-tenant variant of the schema without RBAC to Db2. GitHub [issue-840](https://github.com/IBM/FHIR/issues/840) has been opened to address this. 
```

Author: ok

----------------------------------------------------------------
# Overview and Concepts

> The IBM FHIR schema is managed in code and is applied to a target database using a JDBC connection. This simplifies development of Derby-based unit-tests, allows a common schema definition to be used across multiple database flavors, simplifies deployment in cloud scenarios, and because the data model dependencies are understood by the code, the schema creation process can be parallelized to reduce deployment times. In practice, however, we've found that the number of threads must be limited due to driving contention in the internal catalog tables (in DB2, for example). We have also found that some parallel object create operations cause internal database deadlocks (notably creating foreign key relationships). The implementation contains a retry loop, but may fail if the retry limit is exceeded. Note that these deadlocks are internal to Db2, not the more common kind typically caused by poor application code.

Reviewer 2: feels a tad run-on
```suggestion
The IBM FHIR schema is managed in code and is applied to a target database using a JDBC connection. This simplifies development of Derby-based unit-tests, allows a common schema definition to be used across multiple database flavors, and simplifies deployment in cloud scenarios.
Because the data model dependencies are understood by the code, the schema creation process can be parallelized to reduce deployment times. In practice, however, we've found that the number of threads must be limited due to driving contention in the internal catalog tables (in DB2, for example). We have also found that some parallel object create operations cause internal database deadlocks (notably creating foreign key relationships). The implementation contains a retry loop, but may fail if the retry limit is exceeded. Note that these deadlocks are internal to Db2, not the more common kind typically caused by poor application code.
```

Author: ok

| --------   | ----------|
| DB2        | Multi-tenant. Supports multiple but isolated tenants within one database/schema |
| DB2        | An SPL stored procedure is used to implement the resource storage logic, reducing the number of application server to database round-trips, improving performance |
| DB2        | Uses `FHIR_TS` tablespace for admin tables, and a tablespace per tenant |
| DB2        | `FHIR_TS` is created using a small extent size for efficiency |
| PostgreSQL | Uses a function for the resource persistence logic |
| PostgreSQL | TEXT type used instead of CLOB for large data values |
| Derby      | Resource persistence is implemented at the DAO layer as a sequence of individual statements instead of one procedure call. At a functional level, the process is identical. Simplifies debugging and supports easier unit-test construction. |

----------------------------------------------------------------
# Schema Design - Physical Data Model

The IBM FHIR Server persists resources using tables defined in a data schema (usually FHIRDATA, but this can be anything). Multiple data schemas can exist in the same database, all managed from a single FHIR_ADMIN schema.

> The resource-specific tables are generated by the `FhirResourceTableGroup` class which is called for each value from `FHIRResourceType.ValueSet`.

Reviewer 2: ```suggestion
The resource-specific tables are generated by the `FhirResourceTableGroup` class which is called for each concrete resource type from the `FHIRResourceType.Value` enum.
```

Author: ok

The schema only stores search values relevant to the most recent resource version.

In addition to tables and indexes, each data schema contains the following sequences:

| Sequence Name | Description |
|----------|---------------------------------------------|
| `fhir_sequence` | Provides sequence numbers for the normalized resource ids |
| `fhir_ref_sequence` | Provides sequence numbers for reference tables such as PARAMETER_NAMES, CODE_SYSTEMS and RESOURCE_TYPES |

These sequences should not require updating and should never be altered as this would risk the integrity of existing data.

> By convention, tables are named using the plural form of the data they represent. The following diagram shows the main relationships among tables in the schema. Note that only the search parameter tables for the Patient resource are shown. Each resource type gets its own set of search parameter tables.

Reviewer 2: I expected the diagram here but its actually a couple sections below

Author: I moved it up

## TABLESPACES

The Db2 multi-tenant schema variant uses a tablespace per tenant. All other database implementations use the default tablespace.

----------------------------------------------------------------
## Schema: FHIR_ADMIN

The Schema FHIR_ADMIN is deployed once per database and contains tables used to manage schema version history and tenant access (in the case of a multi-tenant supported configuration). The schema contains 3 tables:

| Object | Notes | Purpose |
| ----- | ----- | ------- |
| VERSION_HISTORY | | Table used to track the schema version number for each of the managed objects in the data schema (usually FHIRDATA) |
| TENANTS | Db2 only | Table for mapping between tenant name and the tenant id used to isolate the tenant's data in the data schema |
| TENANT_KEYS | Db2 only | Authorization keys allocated to a given given. A tenant can have multiple valid keys at any point in time to support rolling key changes (add new before remove old).
| TENANT_SEQUENCE | Db2 only | Sequence used to provide unique mt_id values for allocated tenants |
| SET_TENANT | Db2 only | Stored procedure. The only way for FHIRUSER to set the value of the SV_TENANT_ID session variable.

> 
> 

Reviewer: ```suggestion

```

Author: ok

These table definitions are more completely described in [DB2MultiTenancy.md](DB2MultiTenancy.md). 

> 
> 

Reviewer: ```suggestion

```

Author: ok

## Logical Resources, Resource Versions and Search Parameters

![Physical Schema](physical_schema_V0024.png)

Each logical resource instance such as a Patient, Device or Observation is stored as a row in the `LOGICAL_RESOURCES` table. A corresponding row is also stored in a resource-specific logical resources tables for example: `PATIENT_LOGICAL_RESOURCES`, `DEVICE_LOGICAL_RESOURCES` or `OBSERVATION_LOGICAL_RESOURCES` etc.

> Each time a logical resource is updated a new version is created. Each new version is stored in the resource-type-specific table `xx_RESOURCES` where xx is the resource type name for example: `PATIENT_RESOURCES`, `DEVICE_RESOURCES` or `OBSERVATION_RESOURCES` etc. The resource payload is rendered in JSON, compressed, and stored in the `DATA` column of this table unless payload offloading has been configured, in which case the `DATA` column will be null. Each version is allocated an integer `VERSION_ID` number starting from 1 which increments by 1 for each new version. Row locking (SELECT FOR UPDATE) guarantees there will be no gaps in the version numbers unless a version-specific patient erase custom operation has been invoked. The patient erase operation is the only time rows from the `xx_RESOURCES` table are ever deleted. REST `DELETE` interactions are implemented as _soft_ deletes and create a new version of the resource with the `xx_RESOURCES.IS_DELETED` column value equal to 'Y'. This column is repeated (denormalized) in the `xx_LOGICAL_RESOURCES` table as a performance optimization. Most queries need only non-deleted resources and the query is much faster if the join to the wide `xx_RESOURCES` table can be avoided. See the [Finding and Reading a Resource](#finding-and-reading-a-resource) section for practical examples on how data is accessed in the schema.

Reviewer 2: i felt this was a bit run-on as well (but its probably debatable), so I took a stab at breaking it up
```suggestion
Each time a logical resource is updated a new version is created. Each new version is stored in the resource-type-specific table `xx_RESOURCES` where xx is the resource type name for example: `PATIENT_RESOURCES`, `DEVICE_RESOURCES` or `OBSERVATION_RESOURCES` etc. Unless payload offloading has been configured, the resource payload is rendered in JSON, compressed, and stored in the `DATA` column of this table. If payload offloading is configured, the `DATA` column will be null instead. Each version is allocated an integer `VERSION_ID` number starting from 1 which increments by 1 for each new version. Row locking (SELECT FOR UPDATE) guarantees there will be no gaps in the version numbers unless a version-specific patient erase custom operation has been invoked. The patient erase operation is the only time rows from the `xx_RESOURCES` table are ever deleted. HTTP `DELETE` interactions are implemented as _soft_ deletes and create a new version of the resource with the `xx_RESOURCES.IS_DELETED` column value equal to 'Y'. This column is repeated (denormalized) in the `xx_LOGICAL_RESOURCES` table as a performance optimization. Most queries need only non-deleted resources and the query is much faster if the join to the wide `xx_RESOURCES` table can be avoided. See the [Finding and Reading a Resource](#finding-and-reading-a-resource) section for practical examples on how data is accessed in the schema.
```

Reviewer: Note, it is not the Patient erase operation, it's just the Resource erase operation. It can be used with a Patient.

Reviewer 2: good point.  i updated my suggestion above to remove the word "patient" in front of erase (in both spots).

Author: addressed both

The following table describes the difference among the different key column names. These names were inherited from older schema implementations, which makes them a little harder to grasp initially:

| Column Name | Description |
| ----------- | ----------- |
| LOGICAL_ID  | The application value used to uniquely identify logical resource records. This value is passed by the application and is typically some form of UUID, and aligns with the FHIR `Resource.Id` |
| LOGICAL_RESOURCE_ID | The database assigned primary key for the LOGICAL_RESOURCES table |

> | RESOURCE_ID | The database assigned primary key for the RESOURCES table |

Reviewer 2: i still hate these names, but at least its well documented :-)

| xx_profiles           | Resource parameters | Profile search parameters. Profile is a search parameter (rather than a type of search parameter) so no reference to PARAMETER_NAMES is needed. |
| xx_security           | Resource parameters | Security search parameters. Security is a search parameter (rather than a type of search parameter) so no reference to PARAMETER_NAMES is needed. |
| xx_current_refs       | Resource parameters | Problems, Medications, Allergies, Drug allergies current lists. |
| str_values                 | Whole system parameters | Whole-system search string values. |
| date_values                | Whole system parameters | Whole-system search date values. |
| resource_token_refs        | Whole system parameters | Whole-system search token values. Mapping table to common_token_values. |
| logical_resource_profiles  | Whole system parameters | Whole-system search profile values. |
| logical_resource_tags      | Whole system parameters | Whole-system search tag values.                |

> | logical_resource_security  | Whole system parameters | Whole-system search security values            |

Reviewer: Should the Erase Table be mentioned here? There was a new long running table for persistence offloading. I couldn't find it when I did my local search.

Author: added `erased_resources`

 last_updated        | timestamp without time zone |           | not null | 
 is_deleted          | character(1)                |           | not null | 
 data                | bytea                       |           |          | 
Indexes:
    "patient_resources_pk" PRIMARY KEY, btree (resource_id)
    "patient_resources_prf_in1" UNIQUE, btree (resource_id)
    "idx_patient_resources_lupd" btree (last_updated)
    "idx_patient_resourceslogical_resource_id" btree (logical_resource_id)
```

The first version of a resource is given a `VERSION_ID` value of 1 and each subsequent version increments this value by 1 leaving no gaps.

To optimize certain queries, the `RESOURCE_ID` for the most recent version of a resource is referenced from the `[resourceType]_LOGICAL_RESOURCES` table with the `CURRENT_RESOURCE_ID` column. This is not enforced by a foreign key because the `[resourceType]_LOGICAL_RESOURCES` record is created first and already contains the intended value of the current `RESOURCE_ID` which has been obtained from a sequence. This approach avoids an extra `UPDATE` on `[resourceType]_LOGICAL_RESOURCES` which is expensive during ingestion.

> The IBM FHIR Server uses soft-delete when processing a FHIR `DELETE` interaction. This creates a new version of the resource with a minimal resource data value and the `IS_DELETED` flag = `Y`. Soft-deletes do not delete existing database records. The Patient Erase custom operation can be used to remove all traces of a patient from a database.

Reviewer 2: seems to duplicate the content toward the end of line 89, but maybe its fine

> The IBM FHIR Server uses soft-delete when processing a FHIR `DELETE` interaction. This creates a new version of the resource with a minimal resource data value and the `IS_DELETED` flag = `Y`. Soft-deletes do not delete existing database records. The Patient Erase custom operation can be used to remove all traces of a patient from a database.

Reviewer: ```suggestion
The IBM FHIR Server uses soft-delete when processing a FHIR `DELETE` interaction. This creates a new version of the resource with a minimal resource data value and the `IS_DELETED` flag = `Y`. Soft-deletes do not delete existing database records. The Erase custom operation can be used to remove all traces of a resource from a database.
```

Author: ok

All timestamps stored in the IBM FHIR Server schema are UTC.

For more details on how the resource payload data is stored, see the next section.

## Scanning Resources

> The IBM FHIR Server implements the whole-system `_history` endpoint to fetch resources in the order they were ingested into the system. This endpoint is described detail in the IBM FHIR Server [Conformance Guide](https://ibm.github.io/FHIR/Conformance/#whole-system-history) guide. This service is backed by the `RESOURCE_CHANGE_LOG` table which records the identity of each resource version as it is ingested. The table is indexed on `CHANGE_TSTAMP` which reflects the UTC last-modified timestamp of the version of the resource.

Reviewer 2: ```suggestion
The IBM FHIR Server implements the whole-system `_history` endpoint to fetch resources in the order they were ingested into the system. This endpoint is described in the IBM FHIR Server [Conformance Guide](https://ibm.github.io/FHIR/Conformance/#whole-system-history) guide. This service is backed by the `RESOURCE_CHANGE_LOG` table which records the identity of each resource version as it is ingested. The table is indexed on `CHANGE_TSTAMP` which reflects the UTC last-modified timestamp of the version of the resource.
```

Author: ok

```
fhirdb=> \d fhirdata.resource_change_log
                        Table "fhirdata.resource_change_log"
       Column        |            Type             | Collation | Nullable | Default 
---------------------+-----------------------------+-----------+----------+---------
 resource_id         | bigint                      |           | not null | 
 resource_type_id    | integer                     |           | not null | 
 logical_resource_id | bigint                      |           | not null | 
 change_tstamp       | timestamp without time zone |           | not null | 
 version_id          | integer                     |           | not null | 
 change_type         | character(1)                |           | not null | 
Indexes:
    "resource_change_log_pk" PRIMARY KEY, btree (resource_id)
    "unq_resource_change_log_ctrtri" UNIQUE, btree (change_tstamp, resource_type_id, resource_id)
```

As reflected in the whole-system-history REST API, the `RESOURCE_CHANGE_LOG` table can be scanned in two ways:

1. Ordered by `RESOURCE_ID`, following the natural order of the primary key
2. Ordered by `CHANGE_TSTAMP`, following the timeline of resources as they arrive

> Filtering and ordering based on `RESOURCE_ID` is the simplest approach because there are no duplicates to deal with although care is still required when reading data near the current time (within the window of the maximum transaction timeout time). This is because ids are allocated before the transaction commits. It is therefore possible for records with a smaler `RESOURCE_ID` value to appear after records which have already been committed. This is not a limitation of the IBM FHIR Server, but just a side-effect that is common in systems like this.

Reviewer 2: ```suggestion
Filtering and ordering based on `RESOURCE_ID` is the simplest approach because there are no duplicates to deal with although care is still required when reading data near the current time (within the window of the maximum transaction timeout time). This is because ids are allocated before the transaction commits. It is therefore possible for records with a smaller `RESOURCE_ID` value to appear after records which have already been committed. This is not a limitation of the IBM FHIR Server, but just a side-effect that is common in systems like this.
```

Author: ok

|------------------------|--------------------|-------------|
| [Number](https://www.hl7.org/fhir/r4/search.html#number) | **`<RESOURCE>`_NUMBER_VALUES** | Numerical Search |
| [Date/DateTime](https://www.hl7.org/fhir/r4/search.html#date) | **`<RESOURCE>`_DATE_VALUES** | Date Search |
| [String](https://www.hl7.org/fhir/r4/search.html#string)<br> [Uri](https://www.hl7.org/fhir/r4/search.html#uri)<br> [Reference](https://www.hl7.org/fhir/r4/search.html#reference)| **`<RESOURCE>`_STR_VALUES** | String, URI, and Reference Search |
| [Token](https://www.hl7.org/fhir/r4/search.html#token) | **`<RESOURCE>`_TOKEN_VALUES** | Token Search |
| [Composite](https://www.hl7.org/fhir/r4/search.html#composite) | **`<RESOURCE>`_COMPOSITES** | Composite Search |
| [Quantity](https://www.hl7.org/fhir/r4/search.html#quantity) | **`<RESOURCE>`_QUANTITY_VALUES** | Quantity Search |
| [Positional](https://www.hl7.org/fhir/r4/location.html#positional) | **`<RESOURCE>`_LATLNG_VALUES** | Positional (near) Search |

For instance, for VISIONPRESCRIPTION there are: 

```
VISIONPRESCRIPTION_NUMBER_VALUES
VISIONPRESCRIPTION_DATE_VALUES
VISIONPRESCRIPTION_STR_VALUES

> VISIONPRESCRIPTION_TOKEN_VALUES

Reviewer 2: outdated

Author: thanks - fixed
